Add create to bundles

mpeck opened the pull request:
Adds create to the bundles resource

drapergeek commented:
@mpeck can you also add this to the `fake` client? https://github.com/operable/cog-api-client#cogapihttpclient

drapergeek commented:
Also would you mind adding some tests to both sides please?

mpeck commented:
I was unaware of the fake client bits. Is this what you were asking for @drapergeek? There was a bit of code for bundle creation in the fake client, though it wasn't in the real client. I added some tests for the http client and updated them a bit for the fake client. Bundle is quite a bit more complex than what is presented here in the fake client. A config is uploaded and validated by the server and then the bundle is actually installed. Is that logic something that needs to be duplicated by the fake client? If so, that may need to wait a bit. All of the validation logic lives in spanner. I don't know if we would want to include spanner as a dependency here, but if we do there will be some work to resolve dependency issues.

drapergeek commented:
>Bundle is quite a bit more complex than what is presented here in the fake client. A config is uploaded and validated by the server and then the bundle is actually installed. Is that logic something that needs to be duplicated by the fake client?

It doesn't need to be duplicated but we need some basic validation. Basically what we have done previously is make sure that the "types" match what you woudl need on the server and if that matches, move forward. So if you require an actual "file" upload for the bundle, then if the user gives you an actual file in the `bundle_create` method, you should assume it is correct and create a "bundle" in the fake server.

It's pretty imperative that we have the fake side, that's what all our testing is based on for flywheel, the tests never hit the real API in flywheel.

drapergeek commented:
@mpeck I thought when @imbriaco and I last discussed this, the API was going to take a path that pointed to a file somewhere. Has that changed? It looks like your tests are sending over straight JSON for the bundle.

mpeck commented:
The endpoint can take either a file or straight json. cogctl sends json, so that's what I tested here.

I'll do a check to make sure we have the correct minimal fields and leave at that, if that works for you.

drapergeek commented:
That sounds good.

drapergeek commented:
>The endpoint can take either a file or straight json. cogctl sends json, so that's what I tested here.


Does that mean we can send a URL or the API requires a file?

imbriaco commented:
@drapergeek The URL isn't necessary anymore because the entire bundle is described by the JSON/YAML file now. We're moving away from the first zip-based bundle format to use Docker for distribution of the actual command code.

mpeck commented:
Is this cool to merge?

drapergeek commented on test/unit/cog_api/http/bundles_test.exs:
```diff
@@ -5,6 +5,24 @@ defmodule CogApi.HTTP.BundlesTest do
+        bundle = Client.bundle_create(valid_endpoint, params) |> get_value
+
+        assert present bundle.id
+        assert present bundle.name
```
Minor: can we assert that the name here is correct based on what was given?

drapergeek commented on test/unit/cog_api/fake/bundles_test.exs:
```diff
@@ -79,15 +78,15 @@ defmodule CogApi.Fake.BundlesTest do
         fake_endpoint,
         bundle.id,
-        %{enabled: "false"}
+        %{enabled: "true"}
```
Why did this change?

mpeck replied:
There isn't currently a way to create a bundle that is already enabled. You have to create it and then enable it. Since this test was just checking to see if you could set the enabled flag with a string, I just flipped it instead of doing two updates.

drapergeek replied:
👍

drapergeek commented:
Two small comments, then :shipit:
